cli.go

```diff
@@ -0,0 +1,212 @@
+	listener, err := net.Listen("tcp", ":"+port)
+	if err != nil {
+		return nil, err
+	}
+
+	// Open the default browser to the callback URL.
+	fmt.Fprintf(os.Stderr, "Complete the login via your OIDC provider. Launching browser to:\n\n    %s\n\n\n", authURL)
+	if err := openURL(authURL); err != nil {
+		fmt.Fprintf(os.Stderr, "Error attempting to automatically open browser: '%s'.\nPlease visit the authorization URL manually.", err)
+	}
+
+	// Start local server
+	go func() {
+		err := http.Serve(listener, nil)
+		if err != nil && err != http.ErrServerClosed {
+			doneCh <- loginResp{nil, err}
+		}
+	}()
+
+	// Wait for either the callback to finish or SIGINT to be received
+	select {
+	case s := <-doneCh:
+		return s.secret, s.err
+	case <-sigintCh:
+		return nil, errors.New("interrupted")
+	}
+}
+
+func fetchAuthURL(c *api.Client, role, mount, port string) (string, error) {
+	data := map[string]interface{}{
+		"role":         role,
+		"redirect_uri": fmt.Sprintf("http://localhost:%s/ui/vault/auth/%s/oidc/callback", port, mount),
+	}
+
+	secret, err := c.Logical().Write(fmt.Sprintf("auth/%s/oidc/auth_url", mount), data)
+	if err != nil {
+		return "", err
+	}
+
+	authURL := secret.Data["auth_url"].(string)
+	if authURL == "" {
+		return "", errors.New(fmt.Sprintf("Unable to authorize role %q. Check Vault logs for more information.", role))
+	}
+
+	return authURL, nil
+}
+
+// openURL opens the specified URL in the default browser of the user.
+// Source: https://stackoverflow.com/a/39324149/453290
+func openURL(url string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default: // "linux", "freebsd", "openbsd", "netbsd"
+		cmd = "xdg-open"
+	}
+	args = append(args, url)
+	return exec.Command(cmd, args...).Start()
+}
+
+// parseError converts error from the API into summary and detailed portions.
+// This is used to present a nicer UI by splitting up *known* prefix sentences
+// from the rest of the text. e.g.
+//
+//    "No response from provider. Gateway timeout from upstream proxy."
+//
+// becomes:
+//
+//    "No response from provider.", "Gateway timeout from upstream proxy."
+func parseError(err error) (string, string) {
+	headers := []string{errNoResponse, errLoginFailed, errTokenVerification}
+	summary := "Login error"
+	detail := ""
+
+	re := regexp.MustCompile(`(?s)Errors:.*\* *(.*)`)
+
+	errorParts := re.FindStringSubmatch(err.Error())
+	switch len(errorParts) {
+	case 0:
+		summary = ""
+	case 1:
+		detail = errorParts[0]
+	case 2:
+		for _, h := range headers {
+			if strings.HasPrefix(errorParts[1], h) {
+				summary = h
+				detail = strings.TrimSpace(errorParts[1][len(h):])
+				break
+			}
+		}
+		if detail == "" {
+			detail = errorParts[1]
+		}
+	}
+
+	return summary, detail
+}
+
+// Help method for OIDC cli
+func (h *CLIHandler) Help() string {
+	help := `
+Usage: vault login -method=oidc [CONFIG K=V...]
```

Comment:
Hey! This code is looking really good - the tests pass and the binary builds.

I've been working on manually giving this a spin locally and I'm not quite doing it right yet. Here's what I'm doing so far.
```
$ vault write sys/plugins/catalog/auth/custom-jwt \
    sha256=e0c59922a6a3db23f21e2ff848a372c09bbebe54959857c1ec46f2deb34918cd \
    command="vault-plugin-auth-jwt"
Success! Data written to: sys/plugins/catalog/auth/custom-jwt

$ vault auth enable custom-jwt
Success! Enabled custom-jwt auth method at: custom-jwt/

$ vault login -method=oidc role=engineering
Unknown auth method: oidc. Use "vault auth list" to see the complete list of
auth methods. Additionally, some auth methods are only available via the HTTP
API.

$ vault login -method='custom-jwt' role=engineering
Unknown auth method: custom-jwt. Use "vault auth list" to see the complete
list of auth methods. Additionally, some auth methods are only available via
the HTTP API.
```
Is that happening because I haven't added a role yet?

Reply:
Are you running Vault from the 1.1 beta branch? The recent change there adds the oidc auth method (for which the implementation lives here, making testing change a bit of a nuisance).

Reply:
Oh I'm not! That must be it, I was on master.

Reply:
I still get funny results, though different. After enabling the `custom-jwt` engine as shown in my previous comment, here's what I get.
```
$ vault login -method=oidc role=engineering
Error authenticating: Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/oidc/oidc/auth_url
Code: 404. Errors:

* no handler for route 'auth/oidc/oidc/auth_url'

$ vault login -method=jwt role=engineering
Unknown auth method: jwt. Use "vault auth list" to see the complete list of
auth methods. Additionally, some auth methods are only available via the HTTP
API.

$ vault login -method=custom-jwt role=engineering
Unknown auth method: custom-jwt. Use "vault auth list" to see the complete
list of auth methods. Additionally, some auth methods are only available via
the HTTP API.

$ vault auth list
Path           Type          Accessor                    Description
----           ----          --------                    -----------
custom-jwt/    custom-jwt    auth_custom-jwt_b0fb6e78    n/a
token/         token         auth_token_a318d668         token based credentials
```
Do the instructions here need to be updated in some way or am I just going about things incorrectly?

Reply:
https://github.com/hashicorp/vault/pull/6222

Reply:
Lots of docs to write... In this case since you're not using the default "oidc" path, you'd need to add `-path=custom-jwt`